### DISCUSSION TITLE
segmentation fault

### DISCUSSION THREAD

**Discussion by chuckgriffith2** [shivammathur/homebrew-php#776] [Q&A / Help]
macOS Monterey 12.0.1
Apache/2.4.48
PHP 8.1.2
httpd.conf:
LoadModule php_module /usr/local/opt/php/lib/httpd/modules/libphp.so "*******"
<FilesMatch "\.php$">
    SetHandler application/x-httpd-php
</FilesMatch>

install works
signing for libphp.so works

error on apache startup:
Failed loading /usr/local/opt/php/lib/php/20210902/opcache.so:  dlopen(/usr/local/opt/php/lib/php/20210902/opcache.so, 0x0009): tried: '/usr/local/opt/php/lib/php/20210902/opcache.so' (code signature in <30DBDF5F-9239-3595-AFEC-34F637BCE20F> '/usr/local/Cellar/php/8.1.2/lib/php/20210902/opcache.so' not valid for use in process: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?)), '/usr/lib/opcache.so' (no such file), '/usr/local/Cellar/php/8.1.2/lib/php/20210902/opcache.so' (code signature in <30DBDF5F-9239-3595-AFEC-34F637BCE20F> '/usr/local/Cellar/php/8.1.2/lib/php/20210902/opcache.so' not valid for use in process: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?)), '/usr/lib/opcache.so' (no such file)

tried signing for opcache.so but doesn't work

commented out opcache in php/8.1/conf.d/ext-opcache.ini
apache starts up properly

but as long as php LoadModule is set, random "child pid ##### exit signal Segmentation fault (11)" occur for .html and .jpg files
php works but supporting .jpg, css, etc. files don't

Why would php handler interfere with non php files?

**Comment by shivammathur** [comment]
I'm not sure why it would segfault for non-php files?

But I would recommend using apache from brew instead of using system apache.

```
brew install httpd
```

**Reply by chuckgriffith2** [reply]
do I need to uninstall the system apache to keep them from conflicting?

**Reply by shivammathur** [reply]
Yes, you will have to stop and unload system apache.

Follow the Apache Installation section in this guide. Make sure you select intel processor in the guide at the beginning first.
https://getgrav.org/blog/macos-monterey-apache-multiple-php-versions

**Reply by chuckgriffith2** [reply]
Wow. Thanks from the frozen North.

I was getting tired of testing on the production server.

**Comment by murraycollingwood** [comment]
> macOS Monterey 12.0.1 Apache/2.4.48 PHP 8.1.2 httpd.conf: LoadModule php_module /usr/local/opt/php/lib/httpd/modules/libphp.so "*******" <FilesMatch ".php$"> SetHandler application/x-httpd-php
> 
> tried signing for opcache.so but doesn't work
> 

How do you sign opcache.so ?  I've been looking for some help and haven't been able to find any.

Also, on my setup I found https://test/index.php works okay, but https://test/ give the segmentation fault (despite the fact that I have "DirectoryIndex index.php" entry in the apache vhosts conf file.

**Reply by murraycollingwood** [reply]
So I used the similar instructions to signing libphp.so, seemed to work, but there is no where to specify the certificate authority when loading opcache.so (it's not like the LoadModule syntax).

I'm still getting the segmentation fault.

Did rebuilding httpd using homebrew solve the problem @chuckgriffith2 ??

**Reply by shivammathur** [reply]
Please use apache from Homebrew, instead of the system apache.
You can follow the Apache Installation section in this guide.
https://getgrav.org/blog/macos-monterey-apache-multiple-php-versions

**Reply by murraycollingwood** [reply]
Hi @shivammathur 

I ran the process of brewing httpd, and everything seemed to work okay.
```
==> Installing dependencies for httpd: ca-certificates and pcre
==> Installing httpd dependency: ca-certificates
==> Pouring ca-certificates--2022-03-29.all.bottle.tar.gz
==> Regenerating CA certificate bundle from keychain, this may take a while...
🍺  /opt/homebrew/Cellar/ca-certificates/2022-03-29: 3 files, 211.6KB
==> Installing httpd dependency: pcre
==> Pouring pcre--8.45.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/pcre/8.45: 204 files, 4.6MB
==> Installing httpd
==> Pouring httpd--2.4.53.arm64_monterey.bottle.tar.gz
==> Caveats
DocumentRoot is /opt/homebrew/var/www.
```

And starting up.
```
MacBook-Pro:httpd murraycollingwood$ brew services start httpd
==> Tapping homebrew/services
Cloning into '/opt/homebrew/Library/Taps/homebrew/homebrew-services'...
remote: Enumerating objects: 1954, done.
remote: Counting objects: 100% (461/461), done.
remote: Compressing objects: 100% (192/192), done.
remote: Total 1954 (delta 322), reused 332 (delta 263), pack-reused 1493
Receiving objects: 100% (1954/1954), 539.25 KiB | 684.00 KiB/s, done.
Resolving deltas: 100% (877/877), done.
Tapped 1 command (45 files, 689.3KB).
==> Successfully started `httpd` (label: homebrew.mxcl.httpd)
```

But when I try visiting http://localhost or my site https://sobs I get "This site can't be reached"
<img width="645" alt="image" src="https://user-images.githubusercontent.com/5810216/161868100-05228e21-9a9e-4d3a-bcbb-b63a0e90e9e1.png">

And I can't find any error log
<img width="1535" alt="image" src="https://user-images.githubusercontent.com/5810216/161868191-599fb147-a344-4a56-8115-0629b72e2b7f.png">

The php-fpm.log is an empty file.
I changed the Listen ports to 80 and 443 (so they are the same as my production systems, just with different domains).
I also went through a few LoadModules, turning on socache_shmcb_module, socache_memcache_module, ssl_module, rewrite_module.
I added a php module: 
```
LoadModule php_module /opt/homebrew/opt/php/lib/httpd/modules/libphp.so "Murray Collingwood"
```

I checked the permissions for the var/log/httpd and ensured everybody can write to this directory.

I should at least get some error message, but nothing.  Can you help?

Cheers
Murray

**Reply by murraycollingwood** [reply]
Just saw the trouble shooting tips:

<img width="465" alt="image" src="https://user-images.githubusercontent.com/5810216/161877874-e1ab9c16-5ac1-40bf-9d3d-71074df42fac.png">

<img width="499" alt="image" src="https://user-images.githubusercontent.com/5810216/161877927-96a199a5-3735-4b42-8fb9-f56fee7f9dd1.png">

The error_log didn't exist, so I created it and then ran the tail, and then restarted the server
<img width="672" alt="image" src="https://user-images.githubusercontent.com/5810216/161878069-53e57d8e-072b-41d6-a211-1bfffc1ac4a1.png">

Hang on.... the httpd process seems to be missing!
Is this another signing issue?

Okay - so how do I run this from the command line....
<img width="1224" alt="image" src="https://user-images.githubusercontent.com/5810216/161878634-fde2128f-8c20-4740-b9f5-c2312bb108b1.png">

So - that is looking like my LoadModule libphp line.  
I removed the certificate authority from the LoadModule line - that helped.
Then there were other configuration errors that came up - fixed those.
Finally got it to run and the error log started working... and then the httpd process disappeared from view.

Then these two things happened:
<img width="1115" alt="image" src="https://user-images.githubusercontent.com/5810216/161879853-4c751b91-ea0b-4648-a115-a974c7a26b3b.png">

Lastly I'm getting php on screen:
<img width="339" alt="image" src="https://user-images.githubusercontent.com/5810216/161879922-2fee6034-0cbc-4297-aa24-4cb59411f475.png">

Added
```
<FilesMatch \.php$>
    SetHandler application/x-httpd-php
</FilesMatch>
```

I'm up and running now and so far everything is working.

Do I need to keep running httpd using 'root'?  (From a security perspective I shouldn't be, but on my local development machine it's not a huge risk as my firewall would block all incoming traffic.)

Cheers
Murray
